### DISCUSSION TITLE
How to do a 1-step "flatten" operation on a Component?

### DISCUSSION THREAD

**Discussion by rgiroud** [gdsfactory/gdsfactory#2710] [Q&A]
Hi, 

Here is my problem. Say that you create a Component and add some ComponentReference to it.

obj1 = gf.Component("obj1")
str_part = gf.components.straight(length = 5, width = 2)
curved_part = gf.components.bend_euler(radius=10, angle=45)

ref_str1 = obj1.add_ref(str_part)
ref_curv = obj1.add_ref(curved_part)
ref_str2 = obj1.add_ref(str_part)

ref_curv.connect("o1", ref_str1.ports["o1"])
ref_str2.connect("o1", ref_curv.ports["o2"])

At this step, obj1 contains those basic shapes, which are connected to each other to create a certain path, i.e. they have a specific positioning relative to each other. 

Now, I want to create a new Component, which will contain (among others) a reference to my first obj1. 

obj2 = gf.Component("obj2")
ref_obj1 = obj2.add_ref(obj1)

Is there any way to "transfer" the references from obj1 to obj2? The result would be that instead of having this hierarchy: 
obj2 < obj1< str_part X2, curved_part
I would directly get
obj2 < str_part X2, curved_part
with the relative positioning (and Ports) kept in the process. 

I figure there is an easy way to do this "flatten" operation, because I guess many people want to do something similar, but I haven't found the corresponding function in the documentation or by filtering the discussion with keywords so I'd rather ask it directly, even tho the solution might be very simple.

**Comment by joamatab** [comment]
Hi, you don't need to transfer the references from obj1 to obj2, as obj1 references will keep maintain the underlying references intact when you do `ref_obj1 = obj2.add_ref(obj1)`

**Reply by rgiroud** [reply]
Hi, 

Thanks for your answer. Indeed, if I use KLayout (for exemple) to display obj1 or obj2, the result will be the same. 

But in the tree of references, I’ll have a tree of depth 3, while I would like to have a tree of depth 2 only. 

I’m asking that because I’m working on a project, where I have some functions that generate Components with multiple Ports, and higher level functions that connect those different Components together or with basic Component (like straight or bend_euler), but at the cost of an additional layer in the tree of references. 

Hence my question about a flattening operation…
